File: evaluation/typescript/src/__tests__/rule_evaluator_test.ts
import test from 'ava';
import { Rule } from '../proto/feature/rule_pb';
import { Feature } from '../proto/feature/feature_pb';
import { FixedStrategy, Strategy } from '../proto/feature/strategy_pb';
import { Clause } from '../proto/feature/clause_pb';
import { RuleEvaluator } from '../ruleEvaluator';
import { User } from '../proto/user/user_pb';
import { SegmentUser } from '../proto/feature/segment_pb';
import { createRule, createSegmentUser, createUser, createVariation } from '../modelFactory';

// Define the type for the test cases
interface TestCase {
  user: User;
  expected: Rule | null;
  description: string;
}

const TestRuleList = {
  EQUALS: createRule(
    'rule-id-1',
    'full-name',
    Clause.Operator.EQUALS,
    ['bucketeer project'],
    'variation-A',
  ),
  STARTS_WITH: createRule(
    'rule-id-2',
    'first-name',
    Clause.Operator.STARTS_WITH,
    ['buck'],
    'variation-A',
  ),
  ENDS_WITH: createRule(
    'rule-id-3',
    'last-name',
    Clause.Operator.ENDS_WITH,
    ['ject'],
    'variation-A',
  ),
  SEGMENT: createRule(
    'rule-id-4',
    '',
    Clause.Operator.SEGMENT,
    ['segment-id-1', 'segment-id-2'],
    'variation-B',
  ),
  IN: createRule('rule-id-5', 'email', Clause.Operator.IN, ['bucketeer@gmail.com'], 'variation-B'),
};

function newTestFeature(): Feature {
  const feature = new Feature();
  feature.setId('feature-id');
  feature.setName('test feature');
  feature.setVersion(1);
  feature.setCreatedAt(Math.floor(Date.now() / 1000)); // Unix timestamp in seconds

  const variations = [
    createVariation('variation-A', 'A', 'Variation A', 'Thing does A'),
    createVariation('variation-B', 'B', 'Variation B', 'Thing does B'),
  ];
  feature.setVariationsList(variations);
  const rules = [
    TestRuleList.EQUALS,
    TestRuleList.STARTS_WITH,
    TestRuleList.ENDS_WITH,
    TestRuleList.SEGMENT,
    TestRuleList.IN,
  ];

  feature.setRulesList(rules);

  const fixedStrategy = new FixedStrategy();
  fixedStrategy.setVariation('variation-B');
  const strategy = new Strategy();
  strategy.setType(Strategy.Type.FIXED);
  strategy.setFixedStrategy(fixedStrategy);

  feature.setDefaultStrategy(strategy);

  return feature;
}

function newSegmentUserIDs(): SegmentUser[] {
  return [
    createSegmentUser('user-id-1', 'segment-id-1', SegmentUser.State.INCLUDED),
    createSegmentUser('user-id-1', 'segment-id-2', SegmentUser.State.INCLUDED),
    createSegmentUser('user-id-2', 'segment-id-1', SegmentUser.State.INCLUDED),
    createSegmentUser('user-id-2', 'segment-id-2', SegmentUser.State.INCLUDED),
    createSegmentUser('user-id-3', 'segment-id-1', SegmentUser.State.INCLUDED),
    createSegmentUser('user-id-4', 'segment-id-2', SegmentUser.State.INCLUDED),
  ];
}

// Define test cases
const testcases: TestCase[] = [
  {
    user: createUser('user-id-1', { 'full-name': 'bucketeer project' }),
    expected: TestRuleList.EQUALS,
    description: 'TestRuleList.EQUALS',
  },
  {
    user: createUser('user-id-1', { 'first-name': 'bucketeer' }),
    expected: TestRuleList.STARTS_WITH,
    description: 'TestRuleList.STARTS_WITH',
  },
  {
    user: createUser('user-id-1', { 'last-name': 'project' }),
    expected: TestRuleList.ENDS_WITH,
    description: 'TestRuleList.ENDS_WITH',
  },
  {
    user: createUser('user-id-3', { email: 'bucketeer@gmail.com' }),
    expected: TestRuleList.IN,
    description: 'TestRuleList.IN',
  },
  {
    user: createUser('user-id-1', null),
    expected: TestRuleList.SEGMENT,
    description: 'TestRuleList.SEGMENT - user-id-1',
  },
  {
    user: createUser('user-id-2', null),
    expected: TestRuleList.SEGMENT,
    description: 'TestRuleList.SEGMENT - user-id-2',
  },
  {
    user: createUser('user-id-3', null),
    expected: null,
    description: 'No rule matched - user-id-3',
  },
  {
    user: createUser('user-id-4', null),
    expected: null,
    description: 'No rule matched - user-id-4',
  },
];

// Test cases
testcases.forEach(({ user, expected, description }) => {
  test(description, (t) => {
    const ruleEvaluator = new RuleEvaluator();
    const values = newSegmentUserIDs();
    const actual = ruleEvaluator.evaluate(newTestFeature().getRulesList(), user, values, {});
    t.deepEqual(actual, expected);
  });
});

File: evaluation/go/rule_evaluator_test.go
// Copyright 2025 The Bucketeer Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package evaluation

import (
	"fmt"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"

	ftproto "github.com/bucketeer-io/bucketeer/v2/proto/feature"
	userproto "github.com/bucketeer-io/bucketeer/v2/proto/user"
)

func TestRuleEvaluator(t *testing.T) {
	f := newFeature()
	testcases := []struct {
		user     *userproto.User
		expected *ftproto.Rule
	}{
		{
			user: &userproto.User{
				Id:   "user-id-1",
				Data: map[string]string{"full-name": "bucketeer project"},
			},
			expected: f.Rules[0],
		},
		{
			user: &userproto.User{
				Id:   "user-id-1",
				Data: map[string]string{"first-name": "bucketeer"},
			},
			expected: f.Rules[1],
		},
		{
			user: &userproto.User{
				Id:   "user-id-1",
				Data: map[string]string{"last-name": "project"},
			},
			expected: f.Rules[2],
		},
		{
			user: &userproto.User{
				Id:   "user-id-3",
				Data: map[string]string{"email": "bucketeer@gmail.com"},
			},
			expected: f.Rules[4],
		},
		{
			user: &userproto.User{
				Id:   "user-id-1",
				Data: nil,
			},
			expected: f.Rules[3],
		},
		{
			user: &userproto.User{
				Id:   "user-id-2",
				Data: nil,
			},
			expected: f.Rules[3],
		},
		{
			user: &userproto.User{
				Id:   "user-id-3",
				Data: nil,
			},
			expected: nil,
		},
		{
			user: &userproto.User{
				Id:   "user-id-4",
				Data: nil,
			},
			expected: nil,
		},
	}
	values := newSegmentUserIDs()
	ruleEvaluator := &ruleEvaluator{}
	for i, tc := range testcases {
		des := fmt.Sprintf("index: %d", i)
		actual, _ := ruleEvaluator.Evaluate(f.Rules, tc.user, values, nil)
		assert.Equal(t, tc.expected, actual, des)
	}
}

func newFeature() *ftproto.Feature {
	return &ftproto.Feature{
		Id:        "feature-id",
		Name:      "test feature",
		Version:   1,
		CreatedAt: time.Now().Unix(),
		Variations: []*ftproto.Variation{
			{
				Id:          "variation-A",
				Value:       "A",
				Name:        "Variation A",
				Description: "Thing does A",
			},
			{
				Id:          "variation-B",
				Value:       "B",
				Name:        "Variation B",
				Description: "Thing does B",
			},
		},
		Rules: []*ftproto.Rule{
			{
				Id: "rule-id-1",
				Strategy: &ftproto.Strategy{
					Type: ftproto.Strategy_FIXED,
					FixedStrategy: &ftproto.FixedStrategy{
						Variation: "variation-A",
					},
				},
				Clauses: []*ftproto.Clause{
					{
						Id:        "clause-id-1",
						Attribute: "full-name",
						Operator:  ftproto.Clause_EQUALS,
						Values:    []string{"bucketeer project"},
					},
				},
			},
			{
				Id: "rule-id-2",
				Strategy: &ftproto.Strategy{
					Type: ftproto.Strategy_FIXED,
					FixedStrategy: &ftproto.FixedStrategy{
						Variation: "variation-A",
					},
				},
				Clauses: []*ftproto.Clause{
					{
						Id:        "clause-id-2",
						Attribute: "first-name",
						Operator:  ftproto.Clause_STARTS_WITH,
						Values:    []string{"buck"},
					},
				},
			},
			{
				Id: "rule-id-3",
				Strategy: &ftproto.Strategy{
					Type: ftproto.Strategy_FIXED,
					FixedStrategy: &ftproto.FixedStrategy{
						Variation: "variation-A",
					},
				},
				Clauses: []*ftproto.Clause{
					{
						Id:        "clause-id-3",
						Attribute: "last-name",
						Operator:  ftproto.Clause_ENDS_WITH,
						Values:    []string{"ject"},
					},
				},
			},
			{
				Id: "rule-id-4",
				Strategy: &ftproto.Strategy{
					Type: ftproto.Strategy_FIXED,
					FixedStrategy: &ftproto.FixedStrategy{
						Variation: "variation-B",
					},
				},
				Clauses: []*ftproto.Clause{
					{
						Id:        "clause-id-4",
						Attribute: "",
						Operator:  ftproto.Clause_SEGMENT,
						Values: []string{
							"segment-id-1",
							"segment-id-2",
						},
					},
				},
			},
			{
				Id: "rule-id-5",
				Strategy: &ftproto.Strategy{
					Type: ftproto.Strategy_FIXED,
					FixedStrategy: &ftproto.FixedStrategy{
						Variation: "variation-B",
					},
				},
				Clauses: []*ftproto.Clause{
					{
						Id:        "clause-id-5",
						Attribute: "email",
						Operator:  ftproto.Clause_IN,
						Values:    []string{"bucketeer@gmail.com"},
					},
				},
			},
		},
		DefaultStrategy: &ftproto.Strategy{
			Type: ftproto.Strategy_FIXED,
			FixedStrategy: &ftproto.FixedStrategy{
				Variation: "variation-B",
			},
		},
	}
}

func newSegmentUserIDs() (values []*ftproto.SegmentUser) {
	values = append(values, &ftproto.SegmentUser{
		UserId:    "user-id-1",
		SegmentId: "segment-id-1",
		State:     ftproto.SegmentUser_INCLUDED,
	})
	values = append(values, &ftproto.SegmentUser{
		UserId:    "user-id-1",
		SegmentId: "segment-id-2",
		State:     ftproto.SegmentUser_INCLUDED,
	})
	values = append(values, &ftproto.SegmentUser{
		UserId:    "user-id-2",
		SegmentId: "segment-id-1",
		State:     ftproto.SegmentUser_INCLUDED,
	})
	values = append(values, &ftproto.SegmentUser{
		UserId:    "user-id-2",
		SegmentId: "segment-id-2",
		State:     ftproto.SegmentUser_INCLUDED,
	})
	values = append(values, &ftproto.SegmentUser{
		UserId:    "user-id-3",
		SegmentId: "segment-id-1",
		State:     ftproto.SegmentUser_INCLUDED,
	})
	values = append(values, &ftproto.SegmentUser{
		UserId:    "user-id-4",
		SegmentId: "segment-id-2",
		State:     ftproto.SegmentUser_INCLUDED,
	})
	return values
}
